src/mono/mono/profiler/browser.c

```diff
@@ -29,25 +30,107 @@ struct _MonoProfiler {
 };
 
 static MonoProfiler browser_profiler;
+static int desired_sample_interval_ms;
+static MonoCallSpec callspec;
 
-#ifdef HOST_WASM
+#ifdef HOST_BROWSER
 
-void
-mono_wasm_profiler_enter ();
+typedef struct _ProfilerStackFrame ProfilerStackFrame;
 
-void
-mono_wasm_profiler_leave (MonoMethod *method);
+struct _ProfilerStackFrame {
+	MonoMethod *method;
+	double start;
+	bool should_record;
+	ProfilerStackFrame *next;
+};
+
+static ProfilerStackFrame *profiler_stack;
+static double last_sample_time;
+static int prev_skips_per_period;
+static int skips_per_period;
+static int sample_skip_counter;
+static int stack_depth;
+
+double mono_wasm_profiler_now ();
+void mono_wasm_profiler_record (MonoMethod *method, double start);
+
+static bool should_record_frame (double now)
+{
+	if (sample_skip_counter < skips_per_period) {
+		return FALSE;
+	}
+
+	// timer resolution in non-isolated contexts: 100 microseconds (decimal number)
+	double ms_since_last_sample = now - last_sample_time;
+
+	if (desired_sample_interval_ms > 0 && last_sample_time != 0) {
+		// recalculate ideal number of skips per period
+		double skips_per_ms = ((double)sample_skip_counter) / ms_since_last_sample;
+		double newskips_per_period = (skips_per_ms * ((double)desired_sample_interval_ms));
+		skips_per_period = ((newskips_per_period + ((double)sample_skip_counter) + ((double)prev_skips_per_period)) / 3);
+		prev_skips_per_period = sample_skip_counter;
+	} else {
+		skips_per_period = 0;
+	}
+	last_sample_time = now;
+	sample_skip_counter = 0;
+
+	return TRUE;
+}
 
 static void
 method_enter (MonoProfiler *prof, MonoMethod *method, MonoProfilerCallContext *ctx)
 {
-	mono_wasm_profiler_enter ();
+	sample_skip_counter++;
+	stack_depth++;
+
+	ProfilerStackFrame* frame = g_new0 (ProfilerStackFrame, 1);
+	double now = frame->start = mono_wasm_profiler_now ();
+	frame->should_record = should_record_frame (now);
+	frame->next = profiler_stack;
```

Comment:
this looks like a `previous`/`prev` not a `next`

Reply:
renamed to `parent_frame`